lib/mayaUsd/fileio/translators/translatorBlendShape.cpp

```diff
@@ -0,0 +1,299 @@
+        }
+
+        _AddBlendShape(
+            blendFn,
+            originalShape,
+            parentTransform,
+            originalPoints,
+            pointIndices,
+            blendShapeName,
+            deltaPoints,
+            targetIdx,
+            1.0f);
+
+        // After adding the main blendShape - now add all the in-betweens
+
+        const std::vector<UsdSkelInbetweenShape> inBetweens = blendShape.GetInbetweens();
+        // no in-betweens, can continue to the next blendShape
+        if (inBetweens.empty()) {
+            continue;
+        }
+
+        for (size_t ib = 0; ib < inBetweens.size(); ++ib) {
+            const auto&       currentInBetween = inBetweens[ib];
+            const std::string inBetweenName = currentInBetween.GetAttr().GetName().GetString();
+            float             ibWeight = 0.f;
+            currentInBetween.GetWeight(&ibWeight);
+
+            VtVec3fArray inBetweenDeltaPoints;
+            currentInBetween.GetOffsets(&inBetweenDeltaPoints);
+            if (inBetweenDeltaPoints.empty()) {
+                TF_RUNTIME_ERROR(
+                    "InBetween BlendShape <%s> for mesh <%s> has no delta points.",
+                    inBetweenName.c_str(),
+                    meshPrim.GetPath().GetText());
+                continue;
+            }
+
+            // InBetween are new blendShapes added to the same target index
+            _AddBlendShape(
+                blendFn,
+                originalShape,
+                parentTransform,
+                originalPoints,
+                pointIndices,
+                inBetweenName,
+                inBetweenDeltaPoints,
+                targetIdx,
+                ibWeight);
+        }
+    }
+    return true;
+}
+
+bool UsdMayaTranslatorBlendShape::ReadAnimations(
+    const UsdSkelSkinningQuery&     skinningQuery,
+    const UsdSkelAnimQuery&         animQuery,
+    const UsdMayaPrimReaderContext& context)
+{
+    MStatus status;
+
+    const UsdPrim& skinnedPrim = skinningQuery.GetPrim();
+    VtTokenArray   skinnedBlendShapes;
+    skinningQuery.GetBlendShapeOrder(&skinnedBlendShapes);
+    // mesh doesn't have any blendshapes
+    if (skinnedBlendShapes.empty()) {
+        return true;
+    }
+
+    const VtTokenArray animBlendShapes = animQuery.GetBlendShapeOrder();
+    if (animBlendShapes.empty()) {
+        // there are no animated blendShapes. Nothing to do
+        return true;
+    }
+
+    std::vector<unsigned int> correctBlendShapeIndices;
+    correctBlendShapeIndices.reserve(skinnedBlendShapes.size());
+    // The animation prim can hold blendShapes from several mesh prims.
+    // So it's necessary to find the the indices only for those on the current mesh prim.
+    for (const TfToken& token : skinnedBlendShapes) {
+        const auto iter = std::find(animBlendShapes.begin(), animBlendShapes.end(), token);
+        const auto idx = static_cast<int>(std::distance(animBlendShapes.begin(), iter));
+        if (idx >= 0) {
```

Comment:
I'll be honest and may not understand the code properly. I thought std::find would return end() if not found, but end() - begin() would be greater than zero? Why not compare iter with end()?

Reply:
I'll add comments to try to clarify this a bit further.
Basically, what we're looking for is the index where token was found. So after calling find, we check the distance from the start and cache that number.

Reply:
Yes, but if find returns end(), we should not cache it?